File: cs25backend/src/main/java/edu/dongguk/cs25backend/security/JwtProvider.java
package edu.dongguk.cs25backend.security;


import edu.dongguk.cs25backend.domain.type.UserRole;
import edu.dongguk.cs25backend.exception.CS25Exception;
import edu.dongguk.cs25backend.exception.ErrorCode;
import edu.dongguk.cs25backend.repository.CustomerRepository;
import io.jsonwebtoken.*;
import io.jsonwebtoken.io.Decoders;
import io.jsonwebtoken.security.Keys;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.InitializingBean;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;

import java.security.Key;
import java.util.Date;

@Slf4j
@Component
@RequiredArgsConstructor
public class JwtProvider implements InitializingBean {
    private final CustomerRepository customerRepository;
    @Value("${jwt.secret}")
    private String secretKey;
    @Value("${jwt.accessExpiredMs}")
    private Long accessExpiredMs;
    @Value("${jwt.refreshExpiredMs}")
    private Long refreshExpiredMs;
    private Key key;


    @Override
    public void afterPropertiesSet() throws Exception {
        byte[] keyBytes = Decoders.BASE64.decode(secretKey);
        this.key = Keys.hmacShaKeyFor(keyBytes);
    }

    private String createToken(Long id, UserRole userRole, boolean isAccess) {
        Claims claims = Jwts.claims();
        claims.put("id", id);
        if(isAccess) {
            claims.put("role", userRole);
        }

        return Jwts.builder()
                .setHeaderParam(Header.TYPE, Header.JWT_TYPE)
                .setClaims(claims)
                .setIssuedAt(new Date(System.currentTimeMillis()))
                .setExpiration(new Date(System.currentTimeMillis() + (isAccess ? accessExpiredMs : refreshExpiredMs)))
                .signWith(SignatureAlgorithm.HS256, secretKey)
                .compact();
    }

    public JwtToken createTotalToken(Long id, UserRole userRole) {
        String accessToken = createToken(id, userRole, true);
        String refreshToken = createToken(id, userRole, false);

        return JwtToken.builder()
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .build();
    }

    public String validRefreshToken(HttpServletRequest request) throws JwtException {
        String refreshToken = refineToken(request);
        Claims claims = validateToken(refreshToken);
        UserLoginForm customer = customerRepository.findByIdAndRefreshToken(Long.valueOf(claims.get("id").toString())).get();

        return createToken(customer.getId(), customer.getRole(), true);
    }

    public String getUserId(String token) {
        return Jwts.parserBuilder().setSigningKey(key).build().parseClaimsJws(token).getBody().get("id").toString();
    }

    public Claims validateToken(String token) throws JwtException {
        return Jwts.parserBuilder()
                .setSigningKey(key)
                .build()
                .parseClaimsJws(token)
                .getBody();
    }

    public static String refineToken(HttpServletRequest request) throws JwtException {
        String beforeToken = request.getHeader("Authorization");

        String afterToken = null;
        if (StringUtils.hasText(beforeToken) && beforeToken.startsWith("Bearer ")) {
            afterToken =  beforeToken.substring(7);
        } else {
            throw new CS25Exception(ErrorCode.TOKEN_INVALID_ERROR);
        }
        return afterToken;
    }
}

File: cs25server/src/main/kotlin/edu/dongguk/cs25server/security/JwtProvider.kt
package edu.dongguk.cs25server.security

import edu.dongguk.cs25server.domain.type.UserRole
import edu.dongguk.cs25server.domain.type.UserRole.*
import edu.dongguk.cs25server.exception.ErrorCode
import edu.dongguk.cs25server.exception.GlobalException
import edu.dongguk.cs25server.repository.CustomerRepository
import edu.dongguk.cs25server.repository.HeadquartersRepository
import edu.dongguk.cs25server.repository.ManagerRepository
import io.jsonwebtoken.*
import io.jsonwebtoken.io.Decoders
import io.jsonwebtoken.security.Keys
import jakarta.servlet.http.HttpServletRequest
import org.springframework.beans.factory.InitializingBean
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import org.springframework.util.StringUtils
import java.security.Key
import java.util.*

@Component
class JwtProvider (
    private val customerRepository: CustomerRepository,

    private val managerRepository: ManagerRepository,

    private val headquartersRepository: HeadquartersRepository,

    @Value("\${jwt.secret}")
    private val secretKey: String,

    @Value("\${jwt.accessExpiredMs}")
    private var accessExpiredMs: Long,

    @Value("\${jwt.refreshExpiredMs}")
    private val refreshExpiredMs: Long,

) : InitializingBean {

    private var key: Key? = null

    override fun afterPropertiesSet() {
        this.key = Keys.hmacShaKeyFor(Decoders.BASE64.decode(secretKey))
    }

    fun createToken(id: Long, role: UserRole, isAccess: Boolean) : String {
        val claims = Jwts.claims()
        claims["id"] = id
        if (isAccess) {
            claims["role"] = role
        }

        return Jwts.builder()
            .setHeaderParam(Header.TYPE, Header.JWT_TYPE)
            .setClaims(claims)
            .setIssuedAt(Date(System.currentTimeMillis()))
            .setExpiration(Date(System.currentTimeMillis() + if (isAccess) {accessExpiredMs.toInt()} else {refreshExpiredMs.toInt()}))
            .signWith(key , SignatureAlgorithm.HS256)
            .compact()
    }

    fun createTotalToken(id: Long, userRole: UserRole) = JwtToken(
        accessToken = createToken(id, userRole, true),
        refreshToken = createToken(id, userRole, false))


    fun reissueToken(request: HttpServletRequest, role: UserRole): String {
        val refreshToken = refineToken(request)
        val userId = getUserId(refreshToken)

        val user: UserLoginForm = when (role) {
            CUSTOMER -> customerRepository.findByIdAndRefreshToken(userId, refreshToken)
//            MANAGER -> managerRepository.findByIdAndRefreshToken(userId, refreshToken)
//            HQ -> headquartersRepository.findByIdAndRefreshToken(userId, refreshToken)
            else -> throw GlobalException(ErrorCode.NOT_FOUND_ERROR)
        } ?: throw GlobalException(ErrorCode.NOT_FOUND_ERROR)

        return createToken(user.getId(), user.getRole(), true)
    }

    @Throws(JwtException::class)
    fun getUserId(token: String) = Jwts.parserBuilder()
        .setSigningKey(key)
        .build()
        .parseClaimsJws(token)
        .body["id"].toString().toLong()

    @Throws(JwtException::class)
    fun validateToken(token: String) : Claims = Jwts.parserBuilder()
        .setSigningKey(key)
        .build()
        .parseClaimsJws(token)
        .body

    companion object {
        @Throws(JwtException::class)
        fun refineToken(request: HttpServletRequest): String {
            val beforeToken = request.getHeader("Authorization")
            var afterToken = if (StringUtils.hasText(beforeToken) && beforeToken.startsWith("Bearer ")) {
                beforeToken.substring(7)
            } else {
                throw GlobalException(ErrorCode.TOKEN_INVALID_ERROR)
            }
            return afterToken
        }
    }




}